t/t1092-sparse-checkout-compatibility.sh

```diff
@@ -693,6 +693,11 @@ test_expect_success 'sparse-index is not expanded' '
 	echo >>sparse-index/untracked.txt &&
 	ensure_not_expanded add . &&
 
+	echo "a test change" >> sparse-index/README.md &&
+	ensure_not_expanded diff &&
+	git -C sparse-index add README.md &&
+	ensure_not_expanded diff --staged &&
+
```

Comment:
We might want to additionally check that diff is working correctly by adding more `test_all_match git diff <args>` tests earlier in the script. In particular, it would be good to have a partially-staged change (index, workdir, and HEAD are all different at a given path) both in and out of the sparse-checkout cone. You might need to use `git sparse-checkout set` to allow the `git add` to work, then `git sparse-checkout set` to modify the cone again. I think that even in those "out of the sparse-checkout cone" cases, you will not get an index expansion. My expectations are based on how the diff code was modified for commands like `git status` and `git checkout`.